### ISSUE TITLE
--watch re-runs tests twice if a test file is changed

### ISSUE THREAD

**Issue by BurningDog** [tapjs/tapjs#577]
_Given:_ running `tap -w` inside an npm script
_When:_ I change a test file (`index.spec.js`)
_Then:_ node-tap re-runs the tests, pauses for a second, then (incorrectly) runs them again

_What I want to happen:_ when running node-tap in watch mode, on file change it should only re-run tests once.

### Setup

```
- src
- src/index.js
- tests
- tests/index.spec.js
```

Strangely, the desired behaviour happens when changing the file under test (`src/index.js`) i.e. the tests only re-run once.

#### src/index.js

```js
export function something(value) {
    return ++value;
}
```

#### src/index.spec.js

```js
const tap = require('tap')
const { something } = require('../src/index')

tap.pass('this is fine')

tap.test('this is a nice test', (t) => {
    const actual = something(2)
    tap.equal(actual, 3)
    t.end()
})
```

### Clearing the console between runs

To tack this on here: is there a way to clear the console output between runs?

**Comment by isaacs** [comment]
I can't seem to reproduce this.  Every time I modify either the src or the test file, it only runs one time.  Maybe you're editing while it's running, and that's queuing up another run?  (Which, if so, would be by design.)

Clearing the screen is a good idea, but I think doing it by default might be annoying sometimes.  (I often like to compare output as I'm editing tests.)  I'll add a `cls` command to the repl.

**Comment by isaacs** [comment]
`cls` repl command is in v14.4

**Comment by BurningDog** [comment]
> Maybe you're editing while it's running, and that's queuing up another run? (Which, if so, would be by design.)

I'm not editing while it's running tests. I make a change and save the file in VS Code, switch to the terminal, wait a few seconds...watch the test run, then watch it run again.

This happens if I edit the file in vim too.

If you can't reproduce I must be doing something obviously wrong, but I've set up tap in a new folder with only one `src` file and one `test` file.

Ooh! Found something! If I rename the `tests` folder to `test` then the tests only re-run once on file save. That's odd.

**Comment by isaacs** [comment]
Ok, that's weird.  I can reproduce it with a folder named `tests` plural.  Digging in now.

**Comment by isaacs** [comment]
I had created the folder named "test", just muscle memory I guess.

**Comment by isaacs** [comment]
Ok, so, the issue is this:

NYC automatically excludes the `test/` directory from coverage.  However, it does _not_ automatically exclude `testS/` from coverage.  It also automatically excludes `*.test.js`, but _not_ `*.spec.js`.  This means that when your tests run, the process info index tracks that a file named `/path/to/project/tests/index.spec.js` was run.  It _also_ tracks that the process named `tests/index.spec.js` loaded the file at `/path/to/project/src/index.js`.

The process name (`externalId` in NYC parlance) is the relative path, but the "files" index is the fully resolved path.  So, it ends up watching both `/path/to/project/tests/index.spec.js` (as a covered file) _and_ `tests/index.spec.js` (as a test file).  Since Chokidar was asked to watch both of these paths, it dutifully responds when there's an update, firing an event for both of them.  The first kicks off the test, and the second queues a second test run.

I could fix this by always passing a fully-resolved path to Chokidar, but at least currently, it's depending on getting the non-absolute-path version to resolve when tests change so it can look up in the index and figure out which files were loaded by that test.

I think the better solution here is to not add the tests as fully-resolved `file` elements if they're already there as relative-path counterparts.  The inverse (ie, not adding relative-path `externalId` elements if they're already in as files) would be a problem, since it would mean that we aren't able to work back to the test ids to re-run the tests when source files change.

**Comment by isaacs** [comment]
I misspoke in my last paragraph.  It is actually preferable to only watch with the fully-resolved paths.  A covered test will always be included in its own test run, so there's no worry that it won't get picked up properly.

This is fixed on 14.4.1.  Thanks!

**Comment by BurningDog** [comment]
Awesome - thank you!!!
